File: 2020/python/day2.py
#!/usr/bin/env python3
import re
import sys


def day2(input_file):
    with open(input_file, "r") as f:
        input = f.read()

    print("Day 2: Password Philosophy")
    print(f"\tPart 1: {__part1(input)}")
    print(f"\tPart 2: {__part2(input)}")


pass_re = re.compile(r"(\d+)-(\d+) ([a-z]): (.*)")


def __part1(input):
    valid = 0
    for line in input.splitlines():
        m = pass_re.fullmatch(line)
        if m is None:
            raise Exception(f"Invalid input: {line}")

        (min, max, letter) = (int(m.group(1)), int(m.group(2)), m.group(3))
        count = 0
        for c in m.group(4):
            if c == letter:
                count += 1

        if min <= count <= max:
            valid += 1

    return valid


def __part2(input):
    valid = 0
    for line in input.splitlines():
        m = pass_re.fullmatch(line)
        if m is None:
            raise Exception(f"Invalid input: {line}")

        (l, r, c, pw) = (
            int(m.group(1)) - 1,
            int(m.group(2)) - 1,
            m.group(3),
            m.group(4),
        )
        if (pw[l] == c or pw[r] == c) and pw[l] != pw[r]:
            valid += 1

    return valid


if __name__ == "__main__":
    day2(sys.argv[1])

File: 2020/javascript/day2.js
#!/usr/bin/env node

const fs = require("fs");

const part1 = (input) => {
  let valid = 0;
  input.split("\n").forEach((line) => {
    const caps = line.match(/(\d+)-(\d+) ([a-z]): (.*)/);
    const [min, max, ch, pw] = [
      parseInt(caps[1]),
      parseInt(caps[2]),
      caps[3][0],
      caps[4],
    ];

    let count = 0;
    for (let c of pw) {
      count += c === ch ? 1 : 0;
    }
    if (min <= count && count <= max) {
      valid++;
    }
  });

  return valid;
};

const part2 = (input) => {
  let valid = 0;
  input.split("\n").forEach((line) => {
    const caps = line.match(/(\d+)-(\d+) ([a-z]): (.*)/);
    const [l, r, c, pw] = [
      parseInt(caps[1]) - 1,
      parseInt(caps[2]) - 1,
      caps[3][0],
      caps[4],
    ];

    if ((pw[l] === c || pw[r] === c) && pw[l] !== pw[r]) {
      valid++;
    }
  });

  return valid;
};

const day2 = (inputFile) => {
  const input = fs.readFileSync(inputFile, "utf8");

  console.log("Day 2: Password Philosophy");
  console.log(`\tPart 1: ${part1(input)}`);
  console.log(`\tPart 2: ${part2(input)}`);
};

module.exports = day2;

if (require.main === module) day2(process.argv[2]);
